Add generators for apps/v1 deployments

ayushpateria opened the pull request:
**What this PR does / why we need it**:
- Adds basic generator `DeploymentBasicAppsGeneratorV1` that is used on running `kubectl create deployment`. 
- Changes default basic generator for `create deployment` to `deployment-basic/apps.v1`.
- Don't fall back in case generator is not available for `create deployment`, instead, return an error.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61205

**Special notes for your reviewer**:

**Release note**:
```release-note
Add generators for `apps/v1` deployments.
```

ayushpateria commented:
/assign @liggit

ayushpateria commented:
/assign @liggitt

ayushpateria commented on pkg/kubectl/cmd/run.go:
```diff
@@ -232,17 +231,7 @@ func RunRun(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *c
-			} else {
-				generatorName = cmdutil.RunV1GeneratorName
-			}
+			generatorName = cmdutil.DeploymentAppsV1GeneratorName
```
I've removed the discovery validation here and moved it to `FallbackGeneratorNameIfNecessary`.

ayushpateria commented on pkg/kubectl/cmd/util/factory_client_access.go:
```diff
@@ -596,6 +600,42 @@ func FallbackGeneratorNameIfNecessary(
+		}
+		if !hasResource {
+			warning(cmdErr, DeploymentAppsV1GeneratorName, DeploymentAppsV1Beta1GeneratorName)
+			return FallbackGeneratorNameIfNecessary(DeploymentAppsV1Beta1GeneratorName, discoveryClient, cmdErr)
```
This essentially checks recursively if the older generator can be used. The order is apps/v1 -> apps/v1beta1 -> extensions/v1beta -> v1 ReplicationController

liggitt replied:
shouldn't you wait to print the warning until you know what version you're going to use?

soltysh replied:
Wouldn't be cleaner to just return the generator name `if hasResource` and use `fallthrough` otherwise, printing the warning when `generatorName != <what we're about to return to user>` just like Jordan suggested?

soltysh commented on pkg/kubectl/run.go:
```diff
@@ -209,6 +210,94 @@ func (DeploymentAppsV1Beta1) Generate(genericParams map[string]interface{}) (run
 	return &deployment, nil
 }
 
+type DeploymentAppsV1 struct{}
```
During one of our previous sig-cli calls we discussed the chance of deprecating `kubectl run` so that in the long run it'll only run a pod and nothing else. `kubectl create <resource>` will be the path to go. So I'd prefer not to add anything new to run, the create part is ok.

soltysh commented:
@ayushpateria and please squash your changes into a single commit.
/ok-to-test

ayushpateria commented on pkg/kubectl/run.go:
```diff
@@ -121,94 +122,6 @@ func (DeploymentV1Beta1) Generate(genericParams map[string]interface{}) (runtime
 	return &deployment, nil
 }
 
-type DeploymentAppsV1Beta1 struct{}
```
Removing `DeploymentAppsV1Beta1` because it's not being used anywhere, and we have plans to deprecate `kubectl run`.

liggitt replied:
how do you know it's not being used anywhere? it's usable by specifying the generator name to the run command, right?

ayushpateria replied:
Please look at https://github.com/kubernetes/kubernetes/blob/02a80b6c5e6efd427f834f11ca97e094d4c7fa0a/pkg/kubectl/cmd/util/factory_client_access.go#L554, there's no `DeploymentAppV1Beta1`

ayushpateria replied:
My bad, will revert this.

liggitt replied:
> there's no DeploymentAppV1Beta1

that's because you removed it in this commit :)

https://github.com/kubernetes/kubernetes/pull/61288/files?w=1#diff-61ef000e48b9358fadec0877c58832fbL557

it is used in released versions:

https://github.com/kubernetes/kubernetes/blob/fc32d2f3698e36b93322a3465f63a14e9f0eaead/pkg/kubectl/cmd/util/factory_client_access.go#L517-L526

ayushpateria replied:
Oops! :)

I probably got confused because we don't fall back in case there is no `apps/v1beta1`
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_client_access.go#L559

Will fix this.

ayushpateria replied:
Is that because if the user specified the generator then we don't fall back?

liggitt replied:
> Is that because if the user specifies the generator then we don't fall back?

something like that. they explicitly asked for a specific output, and we probably should not change what they asked for.

ayushpateria replied:
We do the same for deployments - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create_deployment.go#L115
Should I add fall back for `DeploymentAppV1Beta1` for consistency?

ayushpateria commented:
I have refactored a lot of code and made the fallback logic consistent in `create deployment` and `run`. Ideally, if a user is explicitly providing the generator, we should not fall back and print an error if the generator is not available. Currently, however, we would fall-back in case of `create deployment` and on running `run` with `cronjob/v2alpha1`. I have moved all the discovery checks under `FallbackGeneratorNameIfNecessary`, and this function is only called when the user has not provided the generator, if they have, then we don't fall back.

soltysh reviewed (changes requested):
A few nits wrt fallback strategies.

liggitt commented on pkg/kubectl/cmd/run.go:
```diff
@@ -218,48 +215,28 @@ func RunRun(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *c
-	if err != nil {
-		return err
+		// Falling back because the generator was not provided and the default one could be unavailable.
+		generatorNameTemp, err := cmdutil.FallbackGeneratorNameIfNecessary(generatorName, clientset.Discovery(), cmdErr)
```
Really? If I specify I want apps/v1 explicitly, I'd expect that to fail on an older server, just as if I tried to `create -f file-with-appsv1.yaml`

soltysh replied:
That's partially the current behavior, see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run.go#L260-L263

soltysh replied:
I'm fine changing that behavior, though. We'll need to ensure it's consistent everywhere.

ayushpateria replied:
Yeah, it's not consistent right now in run and create deployment. I've made it consistent in this PR. We don't fallback now if the user has explicitly provided the generator, which makes more sense to me.

ayushpateria replied:
Rebased. Yes, `FallbackGeneratorNameIfNecessary` is being used only from `run.go` and `create_deployment.go`, I've made the logic consistent across both the files.

ayushpateria commented:
/retest

ayushpateria committed:
Add basic generator for apps/v1 deployment

BenTheElder commented:
/retest

soltysh reviewed (approved):
/lgtm
/approve no-issue

janetkuo reviewed (commented):
@ayushpateria please also update the generators doc: https://kubernetes.io/docs/reference/kubectl/conventions/#generators

> During one of our previous sig-cli calls we discussed the chance of deprecating kubectl run so that in the long run it'll only run a pod and nothing else. kubectl create <resource> will be the path to go. So I'd prefer not to add anything new to run, the create part is ok.

@soltysh if so we should add `kubectl create <resource>` in the aforementioned doc (and remove the parts about `kubectl run` after it's deprecated).
